File: zig/src/commons.zig
//! Copyright (C) 2026 Lenik <zephyr@bodz.net>
//! SPDX-License-Identifier: AGPL-3.0-or-later
//!
//! Template example: shared helpers live in src/commons.* (not a real module name).

const std = @import("std");

pub fn copyStream(reader: anytype, writer: anytype) !void {
    var buf: [8192]u8 = undefined;
    while (true) {
        const n = try reader.read(&buf);
        if (n == 0) return;
        try writer.writeAll(buf[0..n]);
    }
}

pub fn copyFileToStdout(path: []const u8) !void {
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    const stdout = std.io.getStdOut().writer();
    try copyStream(file.reader(), stdout);
}

test "copyStream roundtrip" {
    const input = "alpha\nbeta\n";
    var in = std.io.fixedBufferStream(input);
    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(std.testing.allocator);
    try copyStream(in.reader(), out.writer(std.testing.allocator));
    try std.testing.expectEqualStrings(input, out.items);
}

File: gcc/src/commons.c
/*
 * Copyright (C) 2026 Lenik <zephyr@bodz.net>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

#define _GNU_SOURCE

#include "commons.h"

#include <stdio.h>

int copy_stream(FILE *in, FILE *out) {
    char buf[COPY_BUF_SIZE];
    size_t n;

    while ((n = fread(buf, 1, sizeof buf, in)) > 0) {
        if (fwrite(buf, 1, n, out) != n) {
            return -1;
        }
    }
    if (ferror(in)) {
        return -1;
    }
    return 0;
}

int copy_file(const char *prog, const char *path) {
    FILE *f = fopen(path, "rb");
    if (!f) {
        fprintf(stderr, "%s: ", prog);
        perror(path);
        return -1;
    }

    int r = copy_stream(f, stdout);
    if (fclose(f) != 0) {
        r = -1;
    }
    if (r != 0) {
        fprintf(stderr, "%s: write error\n", prog);
        return -1;
    }
    return 0;
}
